packages/core/src/state-management/repository.ts

```diff
@@ -199,6 +205,44 @@ export class Repository {
     return this.stateManager.atTime(base$, opts.atTime)
   }
 
+  /**
+   * Applies commit to the existing state
+   *
+   * @param streamId - Stream ID to update
+   * @param commit - Commit data
+   * @param opts - Stream initialization options (request anchor, wait, etc.)
+   */
+  async applyCommit(
+    streamId: StreamID,
+    commit: any,
+    opts: CreateOpts | UpdateOpts
+  ): Promise<RunningState> {
+    const state$ = await this.stateManager.applyCommit(streamId, commit, opts)
+    await this.applyWriteOpts(state$, opts)
+    return state$
+  }
+
+  /**
+   * Apply options relating to authoring a new commit
+   *
+   * @param state$ - Running State
+   * @param opts - Initialization options (request anchor, publish to pubsub, etc.)
+   * @private
+   */
+  async applyWriteOpts(state$: RunningState, opts: CreateOpts | UpdateOpts) {
+    this.stateManager.applyWriteOpts(state$, opts)
+
+    await this.handlePinOpts(state$, opts)
+  }
+
+  async handlePinOpts(state$: RunningState, opts: CommonOpts) {
+    if (opts.pin) {
+      await this.pin(state$)
```

Comment:
Does this only pin if it's not already pinned? Otherwise it could slow things down?

Reply:
It behaves exactly the same as pinning an already-pinned stream does via the public pin API (minus the step of loading the stream state since we already have it here).  So yes, this probably does slow things down if you set it to `true` for a stream that is already pinned.  But that's a bigger problem with our pinning system in general that probably needs to be optimized for all pin requests, regardless of which API they come in from.

Reply:
Right, I asked because I was unsure of how our existing pin api is set up. I remember that there were some things that could be optimized. Maybe we should capture that somewhere? Seems like there's possibly a large performance gain there iirc?

Reply:
Yep good call.  Made https://www.notion.so/threebox/2785876b2d984540a9eeb4832ee0ca81?v=ff8566cdd0d548a299cf0ee24b41527a&p=3d888998788441ea90b826c4612b5937